src/mirage_fs_unix.ml

```diff
@@ -0,0 +1,255 @@
+    (function
+      | Unix.Unix_error (Unix.ENOENT, _, _) ->
+        Lwt.return (Error (`Not_found key))
+      | Unix.Unix_error (e, _, _) ->
+        Lwt.return (Error (`Storage_error (key, Unix.error_message e)))
+      | e -> Lwt.fail e)
+
+let disconnect _ = Lwt.return ()
+
+(* all mkdirs are mkdir -p *)
+let rec create_directory t key =
+  let path = resolve_filename t.base (Mirage_kv.Key.to_string key) in
+  let check_type path =
+    Lwt_unix.LargeFile.stat path >>= fun stat ->
+    match stat.Lwt_unix.LargeFile.st_kind with
+    | Lwt_unix.S_DIR -> Lwt.return (Ok ())
+    | _ -> Lwt.return (Error (`Dictionary_expected key))
+  in
+  if Sys.file_exists path then check_type path
+  else begin
+    create_directory t (Mirage_kv.Key.parent key) >>= function
+    | Error e -> Lwt.return (Error e)
+    | Ok () ->
+      Lwt.catch (fun () ->
+        Lwt_unix.mkdir path 0o755 >>= fun () -> Lwt.return (Ok ())
+      )
+      (function
+        | Unix.Unix_error (e, _, _) -> Lwt.return (Error (`Storage_error (key, Unix.error_message e)))
+        | e -> Lwt.fail e
+      )
+  end
+
+let open_file t key flags =
+  let path = resolve_filename t.base (Mirage_kv.Key.to_string key) in
+  (* create_directory (Filename.dirname path) *)
+  create_directory t (Mirage_kv.Key.parent key) >>= function
+  | Error e -> Lwt.return (Error e)
+  | Ok () ->
+    Lwt.catch (fun () -> Lwt_unix.openfile path flags 0o644 >|= fun fd -> Ok fd)
+      (function
+        | Unix.Unix_error (Unix.ENOSPC, _, _) -> Lwt.return (Error `No_space)
+        | Unix.Unix_error (e, _, _) -> Lwt.return (Error (`Storage_error (key, Unix.error_message e)))
+        | e -> Lwt.fail e)
+
+let file_or_directory {base} path0 =
+  let path = resolve_filename base (Mirage_kv.Key.to_string path0) in
+  Lwt_unix.LargeFile.stat path >|= fun stat ->
+  match stat.Lwt_unix.LargeFile.st_kind with
+  | Lwt_unix.S_DIR -> Ok `Dictionary
+  | Lwt_unix.S_REG -> Ok `Value
+  | _ -> Error (`Storage_error (path0, "not a regular file"))
+
+(* TODO test this *)
+let exists t path0 =
+  Lwt.catch (fun () -> file_or_directory t path0 >|= function
+    | Error e -> Error e
+    | Ok x -> Ok (Some x))
+    (function
+      | Unix.Unix_error (Unix.ENOENT, _, _) -> Lwt.return (Ok None)
+      | Unix.Unix_error (e, _, _) -> Lwt.return (Error (`Storage_error (path0, Unix.error_message e)))
+      | e -> Lwt.fail e)
+
+let last_modified {base} key =
+  let path = resolve_filename base (Mirage_kv.Key.to_string key) in
+  Lwt.catch (fun () ->
+      Lwt_unix.LargeFile.stat path >|= fun stat ->
+      let mtime = stat.Lwt_unix.LargeFile.st_mtime in
+      match Ptime.of_float_s mtime with
+      | None -> Error (`Storage_error (key, "mtime parsing failed"))
+      | Some ts -> Ok (Ptime.Span.to_d_ps (Ptime.to_span ts))
+    )
+    (function
+      | Unix.Unix_error (e, _, _) -> Lwt.return (Error (`Storage_error (key, Unix.error_message e)))
+      | e -> Lwt.fail e)
+
+let batch _ ?retries:_ret _ = assert false
```

Comment:
You can probably assume that `type batch = t` here and use https://github.com/mirage/irmin/blob/master/src/irmin-mem/irmin_mem.ml#L35-L36

Reply:
From the code, we don't fully understand how the `batch` type/operation works in `irmin` yet. We'll use `set` (single write operations) for now, but would love to have a meeting to sketch out the `batch` approach with you in `irmin` and `mirage-kv-mem`.